src/Raven.Server/Documents/Queries/QueriedDocument.cs

```diff
@@ -0,0 +1,133 @@
+﻿using System.Collections.Generic;
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
+using Sparrow.Json;
+
+namespace Raven.Server.Documents.Queries;
+
+public sealed class QueriedDocument : Document
+{
+    private static readonly QueriedDocument ExplicitNullReference = new();
+    private int _refCount = 1;
+    public int RefCount => _refCount;
+
+    // Cloned document is not stored in LRU, that means disposing this object should only decrease reference at original document. 
+    private QueriedDocument _sourceDocument;
+
+    // This is used by document that were loaded additionaly to our document from query.
+    private QueriedDocument _referencingSingle;
+    private List<QueriedDocument> _referencing;
+
+    public bool CanDispose => _refCount == 1;
+    public bool IsCloned;
+
+    public void LinkReferencedDocument(QueriedDocument additionalDocument)
+    {
+        // loaded document doesn't exists
+        if (additionalDocument is null)
+            return;
+
+        // Since most of queries loads only one document
+        // e.g. load originalDoc.ReferenceId as ref
+        // We can avoid creating a list object for single reference and store it in variable.
+        if (_referencingSingle == null)
+        {
+            _referencingSingle = additionalDocument;
+        }
+        else if (ReferenceEquals(_referencingSingle, ExplicitNullReference) == false)
+        {
+            _referencing ??= new();
+            _referencing.Add(_referencingSingle);
+            // ExplicitNullReference means we've multiple references.
+            _referencingSingle = ExplicitNullReference;
```

Comment:
I understand the motivation of having `_referencingSingle` but I have doubts about the usage of `ExplicitNullReference;` as the marker and relaying on `ReferenceEquals()`

Reply:
Changed